File: Tutorials/ObjC/Example10/Example10/ViewController.m
//
//  ViewController.m
//  Example10
//
//  Created by Penthera on 1/31/20.
//  Copyright © 2020 Penthera. All rights reserved.
//

#import "ViewController.h"
#import <VirtuosoClientDownloadEngine/VirtuosoClientDownloadEngine.h>
#import "DemoPlayerViewController.h"

// ---------------------------------------------------------------------------------------------------------
// IMPORTANT:
// The following three values must be initialzied, please contact support@penthera.com to obtain these keys
// ---------------------------------------------------------------------------------------------------------
static NSString* backplaneUrl = @"replace_with_your_backplane_url";
static NSString* publicKey = @"replace_with_your_public_key";
static NSString* privateKey = @"replace_with_your_private_key";

@interface ViewController () <VirtuosoDownloadEngineNotificationsDelegate>
//
// MARK: Instance data
//
@property (nonatomic,strong) VirtuosoAsset *exampleAsset;
@property (nonatomic, strong) VirtuosoDownloadEngineNotificationManager* downloadEngineNotifications;
@property (nonatomic, strong) NSError* error;

//
// MARK: Outlets
//
@property (weak, nonatomic) IBOutlet UILabel *statusLabel;
@property (weak, nonatomic) IBOutlet UIButton *fastPlayBtn;
@property (weak, nonatomic) IBOutlet UIButton *deleteBtn;
@property (weak, nonatomic) IBOutlet UIProgressView *statusProgressBar;

@end

@implementation ViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    // Important:
    // The download engine runs async in background. We can receive status updates by implementing
    // the VirtuosoDownloadEngineNotificationsDelegate methods. As processing progresses, our delegate
    // will receive callbacks allowing us to refresh the UI.
    self.downloadEngineNotifications = [[VirtuosoDownloadEngineNotificationManager alloc]initWithDelegate:self];
    
    self.statusLabel.text = @"Starting Engine...";

    //
    // Enable the Engine
    VirtuosoDownloadEngine.instance.enabled = TRUE;
    
    // Backplane permissions require a unique user-id for the full range of captabilities support to work
    // Production code that needs this will need a unique customer ID.
    // For demonstation purposes only, we use the device name
    NSString* userName = UIDevice.currentDevice.name;
    
    //
    // Create the engine configuration
    VirtuosoEngineConfig* config = [[VirtuosoEngineConfig alloc]initWithUser:userName
                                                                backplaneUrl:backplaneUrl
                                                                   publicKey:publicKey
                                                                  privateKey:privateKey];
    if (!config)
    {
        dispatch_async(dispatch_get_main_queue(), ^{
            UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Setup Required"
                                                                           message:@"Please contact support@penthera.com to setup the backplaneUrl, pulicKey, and privateKey"
                                                                    preferredStyle:UIAlertControllerStyleAlert];
            UIAlertAction* ok = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                exit(0);
            }];
            [alert addAction:ok];
            [self presentViewController:alert animated:TRUE completion:nil];

        });
        return;
    }

    //
    // Start the Engine
    // This method will execute async, the callback will happen on the main-thread.
    [VirtuosoDownloadEngine.instance startup:config startupCallback:^(kVDE_EngineStartupCode status) {
        if (status == kVDE_EngineStartupSuccess) {
            NSLog(@"Startup succeeded.");
        } else {
            NSLog(@"Startup encountered error.");
        }
    }];
}

-(void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
}

// MARK: Click Handlers

- (IBAction)deleteClicked:(UIButton *)sender
{
    sender.enabled = FALSE;  // block reentry
    if (self.exampleAsset)
    {
        self.statusLabel.text = [NSString stringWithFormat:@"Deleting Asset:%@", self.exampleAsset];
        [self.exampleAsset deleteAssetOnComplete:^
         {
             self.statusLabel.text = @"Deleted.";
             [self loadEngineData];
             [self setEnabledAppearance:self.fastPlayBtn enabled:YES];
         }];
    }
}

- (IBAction)fastPlayBtnClicked:(UIButton *)sender
{
    // block reentrancy while we create the asset
 
    [self setEnabledAppearance:self.fastPlayBtn enabled:NO];
    
    // This button does double duty
    
    if (self.exampleAsset && self.exampleAsset.fastPlayReady) {
        // Asset has already been queued for FastPlay and is ready for playback
        
        [self fastPlayAsset:self.exampleAsset];
    }
    else {
        // Queue asset for FastPlay.  When enough of the asset has been processed playblack
        // will automatically begin
        
        if ( kVDE_ReachableViaWiFi != [VirtuosoDownloadEngine.instance currentNetworkStatus] ) {
            UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Network Unreachable" message:@"Demo requires WiFi" preferredStyle:UIAlertControllerStyleAlert];
            [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
            }]];
            [self presentViewController:alert animated:TRUE completion:nil];
            return;
        }
        
        // Important:
        // AssetID should be unique across your video catalog
        NSString* myAssetID = @"tears-of-steel-asset-1";
        
        // Important:
        // Create asset configuration object
        VirtuosoAssetConfig* config = [[VirtuosoAssetConfig alloc]initWithURL:@"http://virtuoso-demo-content.s3.amazonaws.com/tears/index.m3u8"
                                                                      assetID:myAssetID
                                                                  description:@"Tears of Steel"
                                                                         type:kVDE_AssetTypeHLS];
        
        if (!config)
        {
            NSLog(@"create config failed");
            [self setEnabledAppearance:self.fastPlayBtn enabled:YES];
            return;
        }
        
        // FastPlay specific configuration flags
        
        // Skip automatic download
        
        config.autoAddToQueue = false;
        
        // Enable FastPlay
        
        config.fastPlayEnabled = true;
            
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
            
            // Important:
            // Creates the Asset and automatically begins parsing and downloading
            // Status messages are provided via VirtuosoDownloadEngineNotificationsDelegate methods.
            
            [VirtuosoAsset assetWithConfig:config];
            
            dispatch_async(dispatch_get_main_queue(), ^{
                [self refreshView];
                self.statusLabel.text = @"Preparing for FastPlay...";
            });
            
        });
    }
}

//
// MARK: Internal implementation
//

- (void)displayAsset:(VirtuosoAsset*)asset
{
    double fractionComplete = asset.fractionComplete;
    kVDE_DownloadStatusType status = asset.status;
    
    if (asset.fastPlayReady)
    {
        self.statusLabel.text = [NSString stringWithFormat:@"FastPlay ready at %0.02f%% complete (%qi MB)",fractionComplete*100.0,asset.currentSize/1024/1024];
        self.statusProgressBar.progress = fractionComplete;
    }
    else
    {
        if( status != kVDE_DownloadComplete && status != kVDE_DownloadProcessing )
        {
            NSString* statusMessage = (asset.status == kVDE_DownloadInitializing) ? NSLocalizedString(@"Initializing: ", @"") : @"";
            NSString* errorMessage = (asset.downloadRetryCount > 0)? [NSString stringWithFormat:@" (Errors: %i)",asset.downloadRetryCount] : @"";
            
            self.statusLabel.text = [NSString stringWithFormat:@"%@%0.02f%% (%qi MB)%@", statusMessage, fractionComplete*100.0, asset.currentSize/1024/1024, errorMessage];
            self.statusProgressBar.progress = fractionComplete;
            [self.fastPlayBtn setEnabled:NO];
        }
        else if( status == kVDE_DownloadComplete )
        {
            self.statusLabel.text = [NSString stringWithFormat:@"%0.02f%% (%qi MB)",fractionComplete*100.0,asset.currentSize/1024/1024];
            self.statusProgressBar.progress = fractionComplete;
        }
    }
}

-(void)playAsset:(VirtuosoAsset*)asset
{
    [self playAsset:asset playbackType:kVDE_AssetPlaybackTypeLocal];
}

-(void)fastPlayAsset:(VirtuosoAsset*)asset
{
    [self playAsset:asset playbackType:kVDE_AssetPlaybackTypeFastPlay];
}

-(void)playAsset:(VirtuosoAsset*)asset playbackType:(kVDE_AssetPlaybackType)playbackType
{
    [self setEnabledAppearance:self.fastPlayBtn enabled:YES];
    
    [VirtuosoAsset isPlayable:asset callback:^(Boolean playable) {
        if( playable )
        {
            UIViewController* player = [DemoPlayerViewController new];
            self.exampleAsset = asset;
            
            [asset playUsingPlaybackType:playbackType
                               andPlayer:(id<VirtuosoPlayer>)player
                               onSuccess:^ {
                                   // Present the player
                                   [self presentViewController:player animated:YES completion:nil];
                               }
                                  onFail:^ {
                                      self.exampleAsset = nil;
                                      self.error = nil;
                                      NSLog(@"Can't play video!");
                                  }];
        }
    }];

}

// ------------------------------------------------------------------------------------------------------------
// IMPORTANT:
// Loads data from the Download Engine.
// ------------------------------------------------------------------------------------------------------------
- (void)loadEngineData {
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
        //
        // Important:
        // Invoked on background thread to prevent blocking UI updates
        //
        if( [[VirtuosoDownloadEngine instance]started] ) {
            NSArray* downloadComplete = [VirtuosoAsset completedAssetsWithAvailabilityFilter:NO];
            NSArray* downloadPending = [VirtuosoAsset pendingAssetsWithAvailabilityFilter:NO];
            
            //
            // Update UI
            dispatch_async(dispatch_get_main_queue(), ^{
                self.error = nil;
                self.exampleAsset = nil;
                
                if ( nil != downloadComplete && downloadComplete.count > 0 )
                {
                    self.exampleAsset = (VirtuosoAsset *)downloadComplete.firstObject;
                }
                else if (nil != downloadPending && downloadPending.count > 0) {
                    self.exampleAsset = (VirtuosoAsset *)downloadPending.firstObject;
                }
                
                [self refreshView];
            });
        } else {
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 5 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
                [self loadEngineData];
            });
            
        }
        
    });
}

-(void)refreshView
{
    if (nil == self.exampleAsset) {
        
        self.statusLabel.text = @"Ready to start FastPlay";
        self.statusProgressBar.progress = 0;
        
        [self setEnabledAppearance:self.deleteBtn enabled:(nil != self.error ? TRUE : FALSE)];

        return;
    }
   
    self.statusLabel.text = (self.exampleAsset.fastPlayReady ? @"FastPlay ready" : @"Preparing for FastPlay...");
    
    [self setEnabledAppearance:self.deleteBtn enabled:YES];
    
    self.statusProgressBar.progress = self.exampleAsset.fractionComplete;
}

-(void)setEnabledAppearance:(UIControl*)control enabled:(Boolean)enabled
{
    [control setEnabled:enabled];
    [control setBackgroundColor:(enabled ? [UIColor blackColor] : [UIColor lightGrayColor])];
}

//
// MARK: VirtuosoDownloadEngineNotificationsDelegate - required methods ONLY
//

// ------------------------------------------------------------------------------------------------------------
//  Called whenever the Engine starts downloading a VirtuosoAsset object.
// ------------------------------------------------------------------------------------------------------------
- (void)downloadEngineDidStartDownloadingAsset:(VirtuosoAsset * _Nonnull)asset
{
    [self displayAsset:asset];
}

// ------------------------------------------------------------------------------------------------------------
// Called whenever the Engine reports progress for a VirtuosoAsset object
// ------------------------------------------------------------------------------------------------------------
- (void)downloadEngineProgressUpdatedForAsset:(VirtuosoAsset * _Nonnull)asset
{
        [VirtuosoAsset isPlayable:asset callback:^(Boolean playable) {
        if (playable)
        {
            self.exampleAsset = asset;
            [self refreshView];
        }
        [self displayAsset:asset];
    }];
}

// ------------------------------------------------------------------------------------------------------------
// Called when an asset is being processed after background transfer
// ------------------------------------------------------------------------------------------------------------
- (void)downloadEngineProgressUpdatedProcessingForAsset:(VirtuosoAsset * _Nonnull)asset
{
    [self displayAsset:asset];
    [self refreshView];

}

// ------------------------------------------------------------------------------------------------------------
// Called whenever the Engine reports a VirtuosoAsset as complete
// ------------------------------------------------------------------------------------------------------------
- (void)downloadEngineDidFinishDownloadingAsset:(VirtuosoAsset * _Nonnull)asset
{
    [self displayAsset:asset];
    [self refreshView];
    [self loadEngineData];
}

// ------------------------------------------------------------------------------------------------------------
// Called whenever the Engine reports a VirtuosoAsset is ready for FastPlay
// ------------------------------------------------------------------------------------------------------------
-(void)downloadEngineFastPlayAssetReady:(VirtuosoAsset *)asset
{
    NSLog(@"FastPlay asset reported ready to play. Asset: %@", asset);    
    [self fastPlayAsset: asset];
}

// ------------------------------------------------------------------------------------------------------------
// Called whenever the Engine encounters an error
// ------------------------------------------------------------------------------------------------------------
-(void)downloadEngineDidEncounterErrorForAsset:(VirtuosoAsset *)asset
                                         error:(NSError *)error
                                          task:(NSURLSessionTask *)task
                                          data:(NSData *)data
                                    statusCode:(NSNumber *)statusCode {
    
    self.error = error;
    [self displayAsset:asset];
    [self refreshView];
    [self loadEngineData];
}

// ------------------------------------------------------------------------------------------------------------
// Called whenever an asset is added to the Engine
// ------------------------------------------------------------------------------------------------------------
-(void)downloadEngineInternalQueueUpdate {
    [self loadEngineData];
}

-(void)downloadEngineStartupComplete:(Boolean)succeeded {
    [self loadEngineData];
}

@end

File: Tutorials/Swift/Example10/Example10/ViewController.swift
//
//  ViewController.swift
//  Example10
//
//  Created by dev on 2/2/20.
//  Copyright © 2020 Penthera. All rights reserved.
//

import UIKit
import AVKit

class ViewController: UIViewController, VirtuosoDownloadEngineNotificationsDelegate {

    // <-- change these to your settings in production
    let backplaneUrl = "replace_with_your_backplane_url"
    let publicKey = "replace_with_your_public_key"
    let privateKey = "replace_with_your_private_key"

    //
    // MARK: Instance data
    //
    var exampleAsset: VirtuosoAsset?
    var downloadEngineNotifications: VirtuosoDownloadEngineNotificationManager!
    var error: Error?
    
    //
    // MARK: Outlets
    //
    @IBOutlet weak var statusLabel: UILabel!
    @IBOutlet weak var fastPlayBtn: UIButton!
    @IBOutlet weak var deleteBtn: UIButton!
    @IBOutlet weak var statusProgressBar: UIProgressView!
       
    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        // Important:
        // The download engine runs async in background. We can receive status updates by implementing
        // the VirtuosoDownloadEngineNotificationsDelegate methods. As processing progresses, our delegate
        // will receive callbacks allowing us to refresh the UI.
        downloadEngineNotifications = VirtuosoDownloadEngineNotificationManager.init(delegate: self)
        
        self.statusLabel.text = "Engine starting..."
        
        //
        // Enable the Engine
        VirtuosoDownloadEngine.instance().enabled = true;
        
        // Backplane permissions require a unique user-id for the full range of captabilities support to work
        // Production code that needs this will need a unique customer ID.
        // For demonstation purposes only, we use the device name
        let userName = UIDevice.current.name
        
        //
        // Create the engine configuration
        guard let config = VirtuosoEngineConfig(user: userName,
                                                backplaneUrl: self.backplaneUrl,
                                                publicKey: self.publicKey,
                                                privateKey: self.privateKey)
        else
        {
            DispatchQueue.main.async {
                let alert = UIAlertController(title: "Setup Required", message: "Please contact support@penthera.com to setup the backplaneUrl, pulicKey, and privateKey", preferredStyle: .alert)
                alert.addAction(UIAlertAction(title: "OK", style: .default, handler: { (action) in
                    exit(0)
                }))
                self.present(alert, animated: true, completion: nil)
            }
            return
        }

        //
        // Start the Engine
        // This method will execute async, the callback will happen on the main-thread.
        VirtuosoDownloadEngine.instance().startup(config) { (status) in
            if status == .vde_EngineStartupSuccess {
                print("Startup succeeded.")
            } else {
                print("Startup encountered error.")
            }
        }
    }
    
    override func viewWillAppear(_ animated: Bool)
    {
        super.viewWillAppear(animated)
    }
    
    // MARK: Click Handlers
    
    @IBAction func deleteBtnClicked(_ sender: UIButton)
    {
        sender.isEnabled = false   // Block reentry
        
        guard let asset = self.exampleAsset else { return }
        
        statusLabel.text = String(format: "Deleting Asset:%@", asset)
        
        asset.delete(onComplete: {
            self.statusLabel.text = "Deleted.";
            self.loadEngineData()
            self.setEnabledAppearance(self.fastPlayBtn, enabled: true)
        })
    }
        
    @IBAction func fastPlayBtnClicked(_ sender: UIButton)
    {
        // block reentrancy while we create the asset
        
        self.setEnabledAppearance(self.fastPlayBtn, enabled: false)
        
        // This button does double duty
        
        if (self.exampleAsset != nil && self.exampleAsset!.fastPlayReady)
        {
            self.playAsset(self.exampleAsset!);
        }
        else
        {
            // Queue asset for FastPlay.  When enough of the asset has been processed playblack
            // will automatically begin
            
            if .vde_ReachableViaWiFi != VirtuosoDownloadEngine.instance().currentNetworkStatus() {
               let alertView = UIAlertController(title: "Network Unreachable", message: "Demo requires WiFi", preferredStyle: .alert)
               alertView.addAction(UIAlertAction(title: "Ok", style: .cancel, handler: { (action) in
                   sender.isEnabled = true
               }))
               self.present(alertView, animated: true, completion: nil)
               return
            }
           
            // Create the Asset on a background thread
            DispatchQueue.global(qos: .background).async {
                // Important: AssetID should be unique across your video catalog
                let yourAssetID = "tears-of-steel-asset-1"

                // Create asset configuration object
                guard let config = VirtuosoAssetConfig(url: "http://virtuoso-demo-content.s3.amazonaws.com/tears/index.m3u8",
                                                      assetID: yourAssetID,
                                                      description: "Tears of Steel",
                                                      type: kVDE_AssetType.vde_AssetTypeHLS) else {
                                                        print("create config failed")
                                                        self.setEnabledAppearance(self.fastPlayBtn, enabled: true)
                                                        return
                }

                // FastPlay specific configuration flags
                
                // Skip automatic download
                config.autoAddToQueue = false;
                
                // Enable FastPlay
                config.enableFastPlay = true;

               // Create asset and commence downloading.
               let _ = VirtuosoAsset.init(config: config)
                    DispatchQueue.main.async {
                        self.refreshView()
                        self.statusLabel.text = "Preparing for FastPlay...";
               }
            }
        }
    }
    
    //
    // MARK: Internal implementation
    //
    
    func displayAsset(asset: VirtuosoAsset)
    {
        let fractionComplete = asset.fractionComplete
        let status = asset.status
        
        if asset.fastPlayReady
        {
            statusLabel.text = String(format: "FastPlay ready at %0.02f%% complete (%qi MB)", fractionComplete*100.0, asset.currentSize/1024/1024)
            statusProgressBar.progress = Float(fractionComplete)
        }
        else
        {
            var errorAdd = ""
            if( asset.downloadRetryCount > 0 ) {
                errorAdd = String(format: " (Errors: %i)", asset.downloadRetryCount)
            }
            if status != .vde_DownloadComplete && status != .vde_DownloadProcessing {
                let optionalStatus = (asset.status == .vde_DownloadInitializing) ? NSLocalizedString("Initializing: ", comment: "") : ""
                statusLabel.text = String(format: "%@ %@%0.02f%% (%qi MB)%@", asset, optionalStatus, fractionComplete*100.0, asset.currentSize/1024/1024, errorAdd)
                statusProgressBar.progress = Float(fractionComplete)
                self.setEnabledAppearance(fastPlayBtn, enabled: false)
            }
            else if status == .vde_DownloadComplete {
                statusLabel.text = String(format: "%@ %0.02f%% (%qi MB)", asset, fractionComplete*100.0, asset.currentSize/1024/1024)
                statusProgressBar.progress = Float(fractionComplete)
            }
        }
    }
    
    private func playAsset(_ asset: VirtuosoAsset)
    {
        self.setEnabledAppearance(fastPlayBtn, enabled: true);
        
        VirtuosoAsset.isPlayable(asset) { (playable) in
            if !playable {
                return // not playable yet
            }

            switch(asset.type)
            {
                case .vde_AssetTypeHLS, .vde_AssetTypeDASH, .vde_AssetTypeNonSegmented:
                    // Present the player
                    
                    let demoPlayer = DemoPlayerViewController()
                
                    asset.play(using: .vde_AssetPlaybackTypeLocal, andPlayer: demoPlayer as VirtuosoPlayer, onSuccess: {
                    self.present(demoPlayer, animated: true, completion: nil)
                    
                }) {
                    self.exampleAsset = nil
                    self.error = nil
                    print("Can't play video!")
                }
                break
                
            case .vde_AssetTypeHSS:
                print("Playback of HSS supported in another Tutorial")
                break
                
            @unknown default:
                break;
            }
        }
    }
    
    // ------------------------------------------------------------------------------------------------------------
    // IMPORTANT:
    // Loads data from the Download Engine.
    // ------------------------------------------------------------------------------------------------------------
        
    private func loadEngineData()
    {
        self.error = nil

        DispatchQueue.global().async {
            //
            // Important:
            // Invoked on background thread to prevent blocking UI updates
            //
            if VirtuosoDownloadEngine.instance().started {
                let downloadComplete = VirtuosoAsset.completedAssets(withAvailabilityFilter: false)
                let downloadPending = VirtuosoAsset.pendingAssets(withAvailabilityFilter: false)
                
                DispatchQueue.main.async {
                    if downloadComplete.count > 0 {
                        self.exampleAsset = downloadComplete.first as? VirtuosoAsset
                    }
                    else if downloadPending.count > 0 {
                        self.exampleAsset = downloadPending.first as? VirtuosoAsset
                    }
                    else {
                        self.exampleAsset = nil
                    }
                    self.refreshView()
                }
            }
            else {
                DispatchQueue.main.async {
                    self.loadEngineData()
                }
            }
        }
    }
    
    private func refreshView()
    {
        guard let asset = self.exampleAsset else {
            
            self.statusLabel.text = "Ready to start FastPlay";
            self.statusProgressBar.progress = Float(0.0);
            
            self.setEnabledAppearance(deleteBtn, enabled: nil != self.error ? true : false)
            
            return
        }
        
        self.statusLabel.text = (self.exampleAsset!.fastPlayReady ? "FastPlay ready" : "Preparing for FastPlay...");
        
        self.setEnabledAppearance(deleteBtn, enabled: true)
        
        self.statusProgressBar.progress = Float(asset.fractionComplete);
    }
    
    private func setEnabledAppearance(_ control:UIControl, enabled: Bool)
    {
        control.isEnabled = enabled;
        control.backgroundColor = (enabled ? UIColor.black :UIColor.lightGray);
        
        return;
    }

    //
    // MARK: VirtuosoDownloadEngineNotificationsDelegate - required methods ONLY
    //
    
    // --------------------------------------------------------------
    //  Called whenever the Engine starts downloading a VirtuosoAsset object.
    // --------------------------------------------------------------
    func downloadEngineDidStartDownloadingAsset(_ asset: VirtuosoAsset)
    {
        displayAsset(asset: asset)
    }
    
    // --------------------------------------------------------------
    // Called whenever the Engine reports progress for a VirtuosoAsset object
    // --------------------------------------------------------------
    func downloadEngineProgressUpdated(for asset: VirtuosoAsset)
    {
        DispatchQueue.global(qos: .background).async {
            VirtuosoAsset.isPlayable(asset) { (playable) in
                if playable {
                    self.exampleAsset = asset;
                    self.refreshView()
                }
                self.displayAsset(asset: asset)
            }
        }
    }
    
    // --------------------------------------------------------------
    // Called when an asset is being processed after background transfer
    // --------------------------------------------------------------
    func downloadEngineProgressUpdatedProcessing(for asset: VirtuosoAsset)
    {
        displayAsset(asset: asset)
        refreshView()
    }
    
    // --------------------------------------------------------------
    // Called whenever the Engine reports a VirtuosoAsset as complete
    // --------------------------------------------------------------
    func downloadEngineDidFinishDownloadingAsset(_ asset: VirtuosoAsset)
    {
        // Download is complete
        displayAsset(asset: asset)
        refreshView()
        loadEngineData()
    }
    
    // ------------------------------------------------------------------------------------------------------------
    // Called whenever the Engine reports a VirtuosoAsset is ready for FastPlay
    // ------------------------------------------------------------------------------------------------------------
    func downloadEngineFastPlayAssetReady(_ asset: VirtuosoAsset)
    {
        print("FastPlay asset reported ready to play. Asset: %@", asset);
        
        self.playAsset(asset)
    }
    
    // --------------------------------------------------------------
    // Called whenever an Engine encounters error downloading asset
    // --------------------------------------------------------------
    func downloadEngineDidEncounterError(for asset: VirtuosoAsset, error: Error?, task: URLSessionTask?, data: Data?, statusCode: NSNumber?) {
        self.error = error
        displayAsset(asset: asset)
        refreshView()
        loadEngineData()
    }

    // --------------------------------------------------------------
    // Called whenever an asset is added to the Engine
    // --------------------------------------------------------------
    func downloadEngineInternalQueueUpdate() {
        loadEngineData()
    }
    
    // --------------------------------------------------------------
    // Called whenever Engine start completes
    // --------------------------------------------------------------
    func downloadEngineStartupComplete(_ succeeded: Bool) {
        loadEngineData()
    }
}

